Assets/HoloToolkit/UX/Scripts/Buttons/Button.cs

```diff
@@ -0,0 +1,488 @@
+                    m_HandCount--;
+                    m_bHandVisible = m_HandCount <= 0;
+                }
+            }
+        }
+        #endregion
+
+        /// <summary>
+        /// Called when button is pressed down.
+        /// </summary>
+        protected void DoButtonPressed(bool bRelease = false)
+        {
+            ButtonStateEnum newState = ButtonStateEnum.Pressed;
+            this.OnStateChange(newState);
+
+            if (OnButtonPressed != null)
+            {
+                OnButtonPressed(gameObject);
+            }
+
+            if (bRelease)
+            {
+                StartCoroutine("DelayedRelease", 0.2f);
+            }
+        }
+
+        /// <summary>
+        /// Called when button is released.
+        /// </summary>
+        protected void DoButtonReleased()
+        {
+            ButtonStateEnum newState;
+
+            if(m_bFocused)
+            {
+                newState = m_bHandVisible ? ButtonStateEnum.Targeted : ButtonStateEnum.ObservationTargeted;
+            }
+            else
+            {
+                newState = m_bHandVisible ? ButtonStateEnum.Interactive : ButtonStateEnum.Observation;
+            }
+
+            this.OnStateChange(newState);
+
+            if (OnButtonReleased != null)
+            {
+                OnButtonReleased(gameObject);
+            }
+        }
+
+        /// <summary>
+        /// Delayed function to release button works for click events
+        /// </summary>
+        /// <param name="delay"></param>
+        /// <returns></returns>
+        private IEnumerator DelayedRelease(float delay)
+        {
+            yield return new WaitForSeconds(delay);
+            DoButtonReleased();
+        }
+
+        /// <summary>
+        /// Called while button is pressed down.
+        /// </summary>
+        protected void DoButtonHeld()
+        {
+            if (OnButtonHeld != null)
+            {
+                OnButtonHeld(gameObject);
+            }
+        }
+
+        /// <summary>
+        /// Called when something interrupts the button pressed state.
+        /// </summary>
+        protected void DoButtonCancelled()
+        {
+            if (OnButtonCancelled != null)
+            {
+                OnButtonCancelled(gameObject);
+            }
+        }
+
+        /// <summary>
+        /// Use LateUpdate to check for whether or not the hand is up
+        /// </summary>
+        public void LateUpdate()
+        {
+            if (!m_disabled && m_bLastHandVisible != m_bHandVisible)
+            {
+                OnHandVisibleChange(m_bHandVisible);
+            }
+        }
+
+        /// <summary>
+        /// Event to fire off when hand visibity changes
+        /// </summary>
+        /// <param name="visible"></param>
+        public virtual void OnHandVisibleChange(bool visible)
+        {
+            m_bLastHandVisible = visible;
+
+            ButtonStateEnum newState = ButtonState;
+
+            switch (ButtonState)
```

Comment:
Where's the rest of the cases?
What do we do for the default case?

Reply:
The button only cares about OnHandVisibleChange for active states.  Otherwise it doesn't change the state.  What other states would be altered with hand visibility changes?  Also no need for default since we don't want to do anything anyways.